domain/modelmigration/state/state.go

```diff
@@ -66,3 +68,36 @@ FROM model`, ModelInfo{})
 
 	return result.ControllerUUID, nil
 }
+
+// GetAllInstanceIDs returns all instance IDs from the current model as
+// juju/collections set.
+func (s *State) GetAllInstanceIDs(ctx context.Context) (set.Strings, error) {
+
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Errorf("cannot get database to retrieve instance IDs: %w", err)
+	}
+
+	query := `
+SELECT &instanceID.instance_id
+FROM   machine_cloud_instance`
+	queryStmt, err := s.Prepare(query, instanceID{})
+	if err != nil {
+		return nil, errors.Errorf("preparing retrieve all instance IDs statement: %w", err)
+	}
+
+	var result []instanceID
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, queryStmt).GetAll(&result)
+		if err != nil {
+			return errors.Errorf("retrieving all instance IDs: %w", err)
+		}
+		return nil
+	})
+
+	instanceIDs := transform.Slice[instanceID, string](
```

Comment:
Another nit. Up to you if you agree not a blocker for me. This is very expensive IMO it would be quicker to allocate the set to the same size as result and just for loop the results.

Reply:
Yeah I love that suggestion, done in c71afad04b.